File: scripts/loadQuestions.js
// loadQuestionsComplete.js
import { createClient } from "@supabase/supabase-js";
import * as fs from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";
import { config } from "dotenv";

// Configurar dotenv
config();

// Configurar cliente de Supabase
const supabaseUrl = process.env.SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_SERVICE_KEY;

if (!supabaseUrl || !supabaseKey) {
  console.error("Error: Variables de entorno no definidas");
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseKey);
const __dirname = dirname(fileURLToPath(import.meta.url));
const questionBankPath = join(__dirname, "../data/m2_question_bank_completo.json");

async function loadQuestions() {
  try {
    console.log("Leyendo archivo...");
    const rawData = fs.readFileSync(questionBankPath, "utf8");
    const questionBank = JSON.parse(rawData);

    if (!questionBank.preguntas || !Array.isArray(questionBank.preguntas)) {
      console.error("Formato inesperado");
      return;
    }

    console.log(`Total de preguntas: ${questionBank.preguntas.length}`);

    // Crear una pregunta de prueba completa
    const testQuestion = {
      subject: "M2",
      content: "Pregunta de prueba",
      options: [
        { label: "a", text: "Opción A" },
        { label: "b", text: "Opción B" },
        { label: "c", text: "Opción C" },
        { label: "d", text: "Opción D" },
        { label: "e", text: "Opción E" },
      ],
      correct_answer: "a",
      explanation: "Explicación de prueba",
      habilidad: "Aplicar", // Añadido el campo habilidad
      difficulty: 1,
      active: true,
      area_tematica: "Algebra y funciones",
      tema: "Funciones",
      subtema: "Dominio y recorrido",
    };

    console.log("Insertando pregunta de prueba...");
    const { data: testData, error: testError } = await supabase
      .from("questions")
      .insert([testQuestion]);

    if (testError) {
      console.error("Error con pregunta de prueba:", testError.message);
      if (testError.cause) {
        console.error("Causa:", testError.cause);
      }
      console.error("Error completo:", testError);
      console.error("Detalles:", JSON.stringify(testQuestion, null, 2));

      // Si el error es por el campo habilidad, intentar sin él
      if (testError.message.includes("habilidad")) {
        delete testQuestion.habilidad;
        console.log("Intentando sin el campo habilidad...");
        const { data, error } = await supabase
          .from("questions")
          .insert([testQuestion]);

        if (error) {
          console.error("Sigue habiendo error:", error.message);
          return;
        } else {
          console.log("¡Funcionó sin el campo habilidad!");
        }
      } else {
        return;
      }
    } else {
      console.log("Pregunta de prueba insertada correctamente.");
    }

    // Procesar cada pregunta a la vez
    let successCount = 0;

    for (let i = 0; i < questionBank.preguntas.length; i++) {
      const pregunta = questionBank.preguntas[i];

      // Asegurarse de que las alternativas son un array
      const alternativas = Array.isArray(pregunta.alternativas)
        ? pregunta.alternativas
        : [];

      // Asegurarse de que hay 5 alternativas
      while (alternativas.length < 5) {
        alternativas.push(
          `Opción ${String.fromCharCode(65 + alternativas.length)}`
        );
      }

      // Preparar la pregunta con todos los campos necesarios
      const question = {
        subject: "M2",
        content: pregunta.texto || "",
        options: [
          { label: "a", text: alternativas[0] || "" },
          { label: "b", text: alternativas[1] || "" },
          { label: "c", text: alternativas[2] || "" },
          { label: "d", text: alternativas[3] || "" },
          { label: "e", text: alternativas[4] || "" },
        ],
        correct_answer: (pregunta.correcta || "A").toLowerCase(),
        explanation: pregunta.explicacion || "",
        habilidad: "Aplicar", // Añadido el campo habilidad con valor predeterminado
        area_tematica: pregunta.area_tematica || null,
        tema: pregunta.tema || null,
        subtema: pregunta.subtema || null,
        difficulty: pregunta.dificultad || 1,
        active: true,
      };

      try {
        const { error } = await supabase.from("questions").insert([question]);

        if (error) {
          console.error(`Error en pregunta ${i + 1}:`, error.message);

          // Si el problema es el campo habilidad, intentar sin él
          if (error.message.includes("habilidad")) {
            delete question.habilidad;
            const { error: retryError } = await supabase
              .from("questions")
              .insert([question]);

            if (retryError) {
              console.error(
                `Sigue habiendo error en pregunta ${i + 1}:`,
                retryError.message
              );
            } else {
              successCount++;
            }
          }
        } else {
          successCount++;
          if (successCount % 10 === 0) {
            console.log(`Progreso: ${successCount} preguntas insertadas`);
          }
        }
      } catch (e) {
        console.error(`Excepción en pregunta ${i + 1}:`, e.message);
      }

      // Breve pausa entre inserciones
      await new Promise((r) => setTimeout(r, 100));
    }

    console.log(
      `Proceso completado: ${successCount} de ${questionBank.preguntas.length} preguntas insertadas`
    );
  } catch (error) {
    console.error("Error general:", error);
  }
}

loadQuestions();

File: load_questions.py
"""Inserta preguntas PAES en Supabase usando requests.

Uso:
    python load_questions.py
Requiere que SUPABASE_URL, SUPABASE_ANON_KEY y SUPABASE_SERVICE_KEY estén definidas
(en Nuevo_PAES/.env, por ejemplo).
"""
from __future__ import annotations

import json
import os
import sys
import time
from pathlib import Path

import requests
from dotenv import load_dotenv

BASE_DIR = Path(__file__).resolve().parent
DATA_FILE = BASE_DIR / "data" / "m2_question_bank_completo.json"
REQUIRED_ENV = ("SUPABASE_URL", "SUPABASE_ANON_KEY", "SUPABASE_SERVICE_KEY")


def ensure_env() -> dict[str, str]:
    load_dotenv()
    missing = [name for name in REQUIRED_ENV if not os.getenv(name)]
    if missing:
        print(f"Faltan variables de entorno: {', '.join(missing)}", file=sys.stderr)
        sys.exit(1)
    return {name: os.environ[name] for name in REQUIRED_ENV}


def read_questions() -> list[dict]:
    if not DATA_FILE.exists():
        print(f"No se encontró el archivo {DATA_FILE}", file=sys.stderr)
        sys.exit(1)
    with DATA_FILE.open(encoding="utf-8") as fh:
        payload = json.load(fh)
    preguntas = payload.get("preguntas")
    if not isinstance(preguntas, list):
        print("Formato inesperado: se esperaba la clave 'preguntas' con una lista", file=sys.stderr)
        sys.exit(1)
    return preguntas


def normalise_options(raw: list[str]) -> list[dict[str, str]]:
    opciones = raw[:5]
    while len(opciones) < 5:
        opciones.append("")
    labels = ["a", "b", "c", "d", "e"]
    return [{"label": label, "text": opciones[idx] or ""} for idx, label in enumerate(labels)]


def build_payload(pregunta: dict) -> dict:
    alternativas = pregunta.get("alternativas")
    if not isinstance(alternativas, list):
        alternativas = []
    return {
        "subject": "M2",
        "content": pregunta.get("texto", ""),
        "options": normalise_options(alternativas),
        "correct_answer": (pregunta.get("correcta") or "a").lower(),
        "explanation": pregunta.get("explicacion", ""),
        "area_tematica": pregunta.get("area_tematica"),
        "tema": pregunta.get("tema"),
        "subtema": pregunta.get("subtema"),
        "difficulty": pregunta.get("dificultad", 1),
        "active": True,
        "habilidad": pregunta.get("habilidad"),
    }


def insert_questions(env: dict[str, str], preguntas: list[dict]) -> None:
    base_url = env["SUPABASE_URL"].rstrip("/")
    endpoint = f"{base_url}/rest/v1/questions"
    headers = {
        "apikey": env["SUPABASE_ANON_KEY"],
        "Authorization": f"Bearer {env['SUPABASE_SERVICE_KEY']}",
        "Content-Type": "application/json",
        "Prefer": "return=minimal",
    }

    for idx, pregunta in enumerate(preguntas, start=1):
        payload = build_payload(pregunta)
        response = requests.post(endpoint, headers=headers, data=json.dumps(payload))
        if not response.ok:
            print(f"Error en la pregunta {idx} (HTTP {response.status_code}): {response.text}", file=sys.stderr)
            sys.exit(1)
        if idx % 10 == 0:
            print(f"{idx} preguntas insertadas")
        time.sleep(0.05)

    print(f"Inserción completa: {len(preguntas)} preguntas")


def main() -> None:
    env = ensure_env()
    preguntas = read_questions()
    print(f"Leyendo {len(preguntas)} preguntas desde {DATA_FILE}")
    insert_questions(env, preguntas)


if __name__ == "__main__":
    main()
